test/org/javarosa/model/xform/SMSSerializingVisitorTest.java

```diff
@@ -0,0 +1,161 @@
+ * being serialized to the SMS format.
+ */
+public class SMSSerializingVisitorTest {
+    private FormParseInit formParser;
+    private String SMS;
+    private static final String delimiter = " ";
+    private FormEntryController formEntryController;
+    private FormInstance formInstance;
+
+    @Before
+    public void setup() throws IOException {
+        formParser = new FormParseInit();
+        formParser.setFormToParse(new File(PathConst.getTestResourcePath(), "sms_form.xml").toString());
+        formEntryController = formParser.getFormEntryController();
+        formInstance = formEntryController.getModel().getForm().getInstance();
+
+        SMSSerializingVisitor serializer = new SMSSerializingVisitor();
+
+        ByteArrayPayload payload = (ByteArrayPayload) serializer.createSerializedPayload(formInstance);
+
+        SMS = convertToString(payload.getPayloadStream());
+        SMS = SMS.replace("\\", "").replace("\\\\", "\\");
+    }
+
+    @Test
+    public void SmsNotNull() {
+        assertNotNull(SMS);
+    }
+
+    @Test
+    public void testTagsAreBeingSerialized() {
+        for (String tag : getAnswerTags()) {
+            assertTrue(SMS.contains(tag));
+        }
+    }
+
+    /**
+     * Checks to see if all answers for the selected tags exist. This
+     * targets answers of various data types  and structures such as
+     * groups.
+     */
+    @Test
+    public void testTaggedAnswers() {
+        for (String tag : getAnswerTags()) {
+            assertTrue(taggedAnswerExists(tag, SMS));
+        }
+    }
+
+    /***
+     * Checks to see an answer that doesn't have a tag is present .
+     * Placeholder  is an answer bounded to the <maiden_name> tag
+     */
+    @Test
+    public void testAnswerWithoutATag() {
+        assertFalse(SMS.contains("Placeholder"));
+    }
+
+    /***
+     * Checks to see a tag that doesn't have an answer exists within
+     * the SMS.
+     * the CTY tag is attached to  the country  question.
+     */
+    @Test
+    public void testTagWithNoAnswer() {
+        assertFalse(SMS.contains("CTY"));
+    }
+
+    /**
+     * returns a list of all the tags that are bounded to
+     * nodes that have answers.
+     */
+    private static List<String> getAnswerTags() {
+        List<String> tags = new ArrayList<>();
+
+        tags.add("FN");
+        tags.add("LN");
+        tags.add("DOB");
+        tags.add("CN");
+        tags.add("PIC");
+
+        return tags;
+    }
+
+    /**
+     * This function  checks to see if  a  tagged node within the
+     * SMS has a corresponding answer by counting characters until the
+     * delimiter is reached.
+     *
+     * @param tag
+     * @param sms
+     * @return true if  an answer exists after a tag.
+     */
+    private static boolean taggedAnswerExists(String tag, String sms) {
+        if (!sms.contains(tag)) {
+            return false;
+        }
+
+        String answer = "";
```

Comment:
At the risk of optimizing prematurely, consider using StringBuilder here instead of String. Strings are immutable, so a new object must be allocated (and later garbage collected) for every change.
![screen shot 2018-06-18 at 12 35 42 pm](https://user-images.githubusercontent.com/55099/41557999-4f9045ac-72f4-11e8-9929-3013792bf3cb.png)